# **Some papers that some of us found useful**

[Alignment](https://github.com/sidonieB/bioinfo-utils/blob/master/docs/advice/usefull_reads_phylogenomics.md#alignment)
[Tree concordance](https://github.com/sidonieB/bioinfo-utils/blob/master/docs/advice/usefull_reads_phylogenomics.md#tree-concordance)
[Missing data](https://github.com/sidonieB/bioinfo-utils/blob/master/docs/advice/usefull_reads_phylogenomics.md#missing-data)


## **Alignment**

[Paper](https://academic.oup.com/bioinformatics/advance-article/doi/10.1093/bioinformatics/bty448/5026659) looking at a method to remove non-homologous parts from alignments (working mainly on amino-acid sequences and protein-coding nucleotide sequences).  

## **Tree concordance**

[Paper](https://www.biorxiv.org/content/early/2017/08/22/115774) looking for outlier genes.  

[Paper](https://www.biorxiv.org/content/early/2017/08/22/115774) using phyparts to map gene tree concordance on the supermatrix tree.  
This paper also briefly discusses what cutoff value for support in the input trees to use: they use SH-like support <80 as a cutoff.

## **Missing data**

Papers looking at the impact of missing data when using ASTRAL and other species-tree estimation methods [here](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-018-4619-8) and [here]((https://www.ncbi.nlm.nih.gov/pubmed/29029338).
  
[Paper](https://www.ncbi.nlm.nih.gov/pubmed/29029241) showing that filtering out fragmentary sequences before aligning results in less gene tree estimation errors.
